java/ql/lib/semmle/code/java/security/ImproperIntentVerificationQuery.qll

```diff
@@ -0,0 +1,159 @@
+private class RegisterReceiverCall extends MethodAccess {
+  RegisterReceiverCall() {
+    this.getMethod()
+        .getASourceOverriddenMethod*()
+        .hasQualifiedName("android.content", "Context", "registerReceiver")
+  }
+
+  /** Gets the `BroadcastReceiver` argument to this call. */
+  Expr getReceiverArgument() { result = this.getArgument(0) }
+
+  /** Gets the `IntentFilter` argument to this call. */
+  Expr getFilterArgument() { result = this.getArgument(1) }
+}
+
+/** A configuration to detect uses of `registerReceiver` with system intent actions. */
+private class RegisterSystemActionConfig extends DataFlow::Configuration {
+  RegisterSystemActionConfig() { this = "RegisterSystemActionConfig" }
+
+  override predicate isSource(DataFlow::Node node) { node.asExpr() instanceof SystemActionName }
+
+  override predicate isSink(DataFlow::Node node) {
+    exists(RegisterReceiverCall ma | node.asExpr() = ma.getFilterArgument())
+  }
+
+  override predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
+    exists(ConstructorCall cc |
+      cc.getConstructedType().hasQualifiedName("android.content", "IntentFilter") and
+      node1.asExpr() = cc.getArgument(0) and
+      node2.asExpr() = cc
+    )
+    or
+    exists(MethodAccess ma |
+      ma.getMethod().hasQualifiedName("android.content", "IntentFilter", "create") and
+      node1.asExpr() = ma.getArgument(0) and
+      node2.asExpr() = ma
+    )
+    or
+    exists(MethodAccess ma |
+      ma.getMethod().hasQualifiedName("android.content", "IntentFilter", "addAction") and
+      node1.asExpr() = ma.getArgument(0) and
+      node2.(DataFlow::PostUpdateNode).getPreUpdateNode().asExpr() = ma.getQualifier()
+    )
+  }
+}
+
+/** Holds if `rrc` registers a reciever `orm` to recieve the system action `sa` that doesn't verifiy intents it recieves. */
+predicate registeredUnverifiedSystemReceiver(
+  RegisterReceiverCall rrc, UnverifiedOnReceiveMethod orm, SystemActionName sa
+) {
+  exists(RegisterSystemActionConfig conf, ConstructorCall cc |
+    conf.hasFlow(DataFlow::exprNode(sa), DataFlow::exprNode(rrc.getFilterArgument())) and
+    cc.getConstructedType() = orm.getDeclaringType() and
+    DataFlow::localExprFlow(cc, rrc.getReceiverArgument())
+  )
+}
+
+/** Holds if the XML element `rec` declares a reciever `orm` to recieve the system action named `sa` that doesn't verifiy intents it recieves. */
+predicate xmlUnverifiedSystemReceiver(
+  XMLElement rec, UnverifiedOnReceiveMethod orm, SystemActionName sa
+) {
+  exists(XMLElement filter, XMLElement action, Class ormty |
+    rec.hasName("receiver") and
+    filter.hasName("intent-filter") and
+    action.hasName("action") and
+    filter = rec.getAChild() and
+    action = filter.getAChild() and
+    ormty = orm.getDeclaringType() and
+    rec.getAttribute("name").getValue() = ["." + ormty.getName(), ormty.getQualifiedName()] and
+    action.getAttribute("name") = sa
+  )
+}
```

Comment:
This can be private. Also, I think you can make use of `AndroidManifest.qll` classes:

```suggestion
private predicate xmlUnverifiedSystemReceiver(
  AndroidReceiverXmlElement rec, UnverifiedOnReceiveMethod orm, SystemActionName sa
) {
  exists(Class ormty |
    ormty = orm.getDeclaringType() and
    rec.getComponentName() = ["." + ormty.getName(), ormty.getQualifiedName()] and
    rec.getAnIntentFilterElement().getAnActionElement().getActionName() = sa
  )
}
```